src/ServiceControl.Config/Commands/UpgradeAuditInstanceCommand.cs

```diff
@@ -0,0 +1,125 @@
+
+    class UpgradeAuditInstanceCommand : AwaitableAbstractCommand<InstanceDetailsViewModel>
+    {
+        readonly IEventAggregator eventAggregator;
+        readonly IWindowManagerEx windowManager;
+        readonly ServiceControlInstanceInstaller serviceControlInstaller;
+        readonly ServiceControlAuditInstanceInstaller serviceControlAuditInstaller;
+        readonly Func<string, AddNewAuditInstanceViewModel> auditUpgradeViewModelFactory;
+
+        [FeatureToggle(Feature.LicenseChecks)]
+        public bool LicenseChecks { get; set; }
+
+        public UpgradeAuditInstanceCommand(Func<InstanceDetailsViewModel, bool> canExecuteMethod = null) : base(canExecuteMethod)
+        {
+        }
+
+        public UpgradeAuditInstanceCommand(
+            IWindowManagerEx windowManager,
+            IEventAggregator eventAggregator,
+            ServiceControlInstanceInstaller serviceControlInstaller,
+            ServiceControlAuditInstanceInstaller serviceControlAuditInstaller,
+            Func<string, AddNewAuditInstanceViewModel> auditUpgradeViewModelFactory)
+        {
+            this.windowManager = windowManager;
+            this.eventAggregator = eventAggregator;
+            this.serviceControlInstaller = serviceControlInstaller;
+            this.serviceControlAuditInstaller = serviceControlAuditInstaller;
+            this.auditUpgradeViewModelFactory = auditUpgradeViewModelFactory;
+        }
+
+        public async override Task ExecuteAsync(InstanceDetailsViewModel model)
+        {
+            if (LicenseChecks)
+            {
+                var licenseCheckResult = serviceControlInstaller.CheckLicenseIsValid();
+                if (!licenseCheckResult.Valid)
+                {
+                    windowManager.ShowMessage("LICENSE ERROR", $"Upgrade could not continue due to an issue with the current license. {licenseCheckResult.Message}.  Contact contact@particular.net", hideCancel: true);
+                    return;
+                }
+            }
+
+            var instance = InstanceFinder.FindInstanceByName<ServiceControlAuditInstance>(model.Name);
+            instance.Service.Refresh();
+
+            var upgradeInfo = UpgradeControl.GetUpgradeInfoForTargetVersion(serviceControlInstaller.ZipInfo.Version, instance.Version);
+            var upgradeOptions = new ServiceControlUpgradeOptions { UpgradeInfo = upgradeInfo };
+
+            if (instance.Service.Status != ServiceControllerStatus.Stopped &&
+                !windowManager.ShowYesNoDialog($"STOP INSTANCE AND UPGRADE TO {serviceControlInstaller.ZipInfo.Version}",
+                    $"{model.Name} needs to be stopped in order to upgrade to version {serviceControlInstaller.ZipInfo.Version}.",
+                    "Do you want to proceed?",
+                    "Yes I want to proceed", "No"))
+            {
+                return;
+            }
+            
+            await UpgradeAuditInstance(model, instance, upgradeOptions);
```

Comment:
It is called on the UI. AFAIK you don't need it if the caller code is not blocking (e.g. not calling `ExecuteAsync.Result`). We don't block any async methods across the SC codebase.

Reply:
This is called on the UI. You have to be super careful to do `ConfigureAwait(false)` there because the code that comes in the continuation can then no longer marshal back to the UI thread